fix(mobile): #15182 Video memories no longer play

Tyris opened the pull request:
I believe this was introduced along with the introduction of the currentAssetRepository.

The fix (for #15182) is relatively simple by just ensuring the right currentAsset gets set on memoryPage load and transition.

Unfortunately, while this fixes the issue, I now have pretty weird stutter that shows for a few hundred ms between the preview image and playing the actual video on my Pixel 8. There is already some timers to try to prevent "stutters" (which may be related), but I found I had to increase them significantly (300ms to ~1s) to hide the issue completely. I'm guessing it's related to funniness in Android's video surface (which is known to be a bit dodgy); but I tried messing around with waiting till the video was initialised/loading/playing, but couldn't get it going.

Since I'm not sure if the above stuttering is a regression (I'm very new to using the app) - I thought I'd raise this PR but ask someone who's more familiar with this area of logic to have a look.

tomgquest committed:
Update current asset to play video.

mertalev commented on mobile/lib/pages/photos/memory.page.dart:
```diff
@@ -125,8 +128,15 @@ class MemoryPage extends HookConsumerWidget {
     if (currentAssetPage.value == 0) {
+      Future.delayed(const Duration(milliseconds: 1)).then((_) {
+        final asset = currentMemory.value.assets[0];
+        ref.read(currentAssetProvider.notifier).set(asset);
```
The current asset should be set *before* this page is being built. Setting it in a future on the page itself is hacky.

Tyris replied:
I agree, but where would you recommend? This is really the earliest place it can be done that makes sense since we don't know the asset prior to this point (I believe - I'm still getting used to the code-base and Flutter).

I tried doing it in a useEffect, but Flutter doesn't like that. I used a future because we do the same already for caching the "next" asset (and Flutter actually suggested I use a Future when I tried it in useEffect).

mertalev replied:
You can set it before pushing the route [here](https://github.com/immich-app/immich/blob/7d50d3032bff9c32c1d001ce40ac184444009d5b/mobile/lib/widgets/memories/memory_lane.dart#L34)

mertalev replied:
It's what the Riverpod dev [recommends](https://github.com/rrousselGit/flutter_hooks/issues/153#issuecomment-997423464); it avoids rebuilding the widget one frame after it's built and means the widget can assume the current asset is set without having to bootstrap itself.

mertalev commented:
> There is already some timers to try to prevent "stutters" (which may be related), but I found I had to increase them significantly (300ms to ~1s) to hide the issue completely. I'm guessing it's related to funniness in Android's video surface (which is known to be a bit dodgy); but I tried messing around with waiting till the video was initialised/loading/playing, but couldn't get it going.

It depends on how long the animation takes. The video initialization kind of competes with Flutter animations, so the timers are to make sure the video waits for the animation. The memory animation might be slow enough that the existing timers aren't enough. Other animations like swiping and maybe the hero animation were made faster so the video can wait for them without waiting too long.

tomgquest committed:
Updated location of currentAssetProvider update per feedback.

tomgquest committed:
Added a playbackDelayFactor to the video viewer to resolve an issue in memories.

Also adjusted the scale of the memory preview image to match the ratio of the video. This still appears to jump because the video preview doesn't seem to be the first frame for some reason :\

Tyris commented:
> > There is already some timers to try to prevent "stutters" (which may be related), but I found I had to increase them significantly (300ms to ~1s) to hide the issue completely. I'm guessing it's related to funniness in Android's video surface (which is known to be a bit dodgy); but I tried messing around with waiting till the video was initialised/loading/playing, but couldn't get it going.
> 
> It depends on how long the animation takes. The video initialization kind of competes with Flutter animations, so the timers are to make sure the video waits for the animation. The memory animation might be slow enough that the existing timers aren't enough. Other animations like swiping and maybe the hero animation were made faster so the video can wait for them without waiting too long.

Have updated the PR with a cludgy playbackDelayFactor to improve this. I noticed that for remote videos, the preview frame is not the first frame, so it looks jumpy when it goes to play - imo it looks better just using the blurhash (removing the preview image entirely) - but open to input.

alextran1502 committed:
Merge branch 'main' into fix/15182-video-memories-no-longer-play

alextran1502 committed:
Merge branch 'main' into fix/15182-video-memories-no-longer-play

alextran1502 committed:
add video indicator

alextran1502 reviewed (approved):
Thank you, I added a video indicator
